File: leetcode/AIcoding/3.DeepSeek-v3/go_solutions/1307-verbal-arithmetic-puzzle.go
func isSolvable(words []string, result string) bool {
    allWords := append([]string{}, words...)
    allWords = append(allWords, result)
    uniqueChars := make(map[byte]bool)
    firstChars := make(map[byte]bool)

    for _, word := range allWords {
        for i := 0; i < len(word); i++ {
            c := word[i]
            uniqueChars[c] = true
            if i == 0 && len(word) > 1 {
                firstChars[c] = true
            }
        }
    }

    chars := make([]byte, 0, len(uniqueChars))
    for c := range uniqueChars {
        chars = append(chars, c)
    }

    n := len(chars)
    if n > 10 {
        return false
    }

    perm := make([]int, 10)
    used := make([]bool, 10)

    var backtrack func(int) bool
    backtrack = func(pos int) bool {
        if pos == n {
            return check(words, result, chars, perm)
        }

        for d := 0; d < 10; d++ {
            if !used[d] {
                if d == 0 {
                    c := chars[pos]
                    if firstChars[c] {
                        continue
                    }
                }
                perm[pos] = d
                used[d] = true
                if backtrack(pos + 1) {
                    return true
                }
                used[d] = false
            }
        }
        return false
    }

    return backtrack(0)
}

func check(words []string, result string, chars []byte, perm []int) bool {
    charToDigit := make(map[byte]int)
    for i, c := range chars {
        charToDigit[c] = perm[i]
    }

    sum := 0
    for _, word := range words {
        num := 0
        for _, c := range word {
            num = num * 10 + charToDigit[byte(c)]
        }
        sum += num
    }

    resNum := 0
    for _, c := range result {
        resNum = resNum * 10 + charToDigit[byte(c)]
    }

    return sum == resNum
}

File: leetcode/AIcoding/3.DeepSeek-v3/python_solutions/1307-verbal-arithmetic-puzzle.py
class Solution:
    def isSolvable(self, words: List[str], result: str) -> bool:
        from collections import defaultdict

        all_words = words + [result]
        first_chars = {word[0] for word in all_words if len(word) > 1}
        n = max(len(word) for word in all_words)
        if len(result) < n:
            return False

        char_to_digit = {}
        digit_to_char = {}

        def backtrack(pos, word_idx, carry, used_digits):
            if pos == n:
                return carry == 0

            if word_idx == len(all_words):
                sum_val = carry
                for word in words:
                    if pos < len(word):
                        sum_val += char_to_digit[word[-(pos + 1)]]
                carry, digit = divmod(sum_val, 10)
                if result[-(pos + 1)] in char_to_digit:
                    if char_to_digit[result[-(pos + 1)]] != digit:
                        return False
                else:
                    if digit in digit_to_char or (digit == 0 and result[-(pos + 1)] in first_chars):
                        return False
                    char_to_digit[result[-(pos + 1)]] = digit
                    digit_to_char[digit] = result[-(pos + 1)]
                    used = used_digits | {digit}
                res = backtrack(pos + 1, 0, carry, used)
                if result[-(pos + 1)] not in char_to_digit:
                    del digit_to_char[digit]
                    del char_to_digit[result[-(pos + 1)]]
                return res

            word = all_words[word_idx]
            if pos >= len(word):
                return backtrack(pos, word_idx + 1, carry, used_digits)
            c = word[-(pos + 1)]
            if c in char_to_digit:
                return backtrack(pos, word_idx + 1, carry, used_digits)
            for d in range(10):
                if d in digit_to_char:
                    continue
                if d == 0 and c in first_chars:
                    continue
                char_to_digit[c] = d
                digit_to_char[d] = c
                if backtrack(pos, word_idx + 1, carry, used_digits | {d}):
                    return True
                del char_to_digit[c]
                del digit_to_char[d]
            return False

        return backtrack(0, 0, 0, set())